nall/gdb/server.cpp

```diff
@@ -0,0 +1,535 @@
+    u32 maxLoopResets = 10000; // how many times can a new message reset the counter (prevents infinite loops with misbehaving clients)
+    bool wasHalted = isHalted();
+
+    for(u32 frame=0; frame<loopFrames; ++frame) {
+      for(u32 i=0; i<loopCount; ++i) {
+        messageCount = 0;
+        update();
+
+        // if the last message resumed the program, abort (no more messages will be send until the next stop)
+        if(wasHalted && !isHalted())return;
+
+        if(messageCount > 0 && maxLoopResets > 0) {
+          i = loopCount; // reset loop here to keep a fast chain of messages going (reduces latency)
+          --maxLoopResets;
+        }
+      }
+      
+      if(wasHalted)usleep(1);
+    }
+  }
+
+  auto Server::getStatusText(u32 port, bool useIPv4) -> string {
+    auto url = getURL(port, useIPv4);
+    string prefix = isHalted() ? "⬛" : "▶";
+
+    if(hasClient())return {prefix, " GDB connected ", url};
+    if(isStarted())return {"GDB listening ", url};
+    return {"GDB pending (", url, ")"};
+  }
+
+  auto Server::sendSignal(Signal code) -> void {
+    sendPayload({"S", hex(static_cast<u8>(code), 2)});
+  }
+
+  auto Server::sendSignal(Signal code, const string& reason) -> void {
+    sendPayload({"T", hex(static_cast<u8>(code), 2), reason});
+  }
+
+  auto Server::sendPayload(const string& payload) -> void {
+    string msg{"+$", payload, '#', hex(gdbCalcChecksum(payload), 2, '0')};
+    if constexpr(GDB_LOG_MESSAGES) {
+      printf("GDB >: %.*s\n", msg.size() > 100 ? 100 : msg.size(), msg.data());
+    }
+    sendText(msg);
+  }
+
+  auto Server::haltProgram() -> void {
+    forceHalt = true;
+    haltSignalSent = false;
+  }
+
+  auto Server::resumeProgram() -> void {
+    pcOverride.reset();
+    forceHalt = false;
+    haltSignalSent = false;
+  }
+
+  auto Server::onConnect() -> void {
+    printf("GDB: TCP connect\n");
+    resetClientData();
+    hasActiveClient = true;
+  }
+
+  auto Server::onDisconnect() -> void {
+    printf("GDB: TCP disconnected\n");
+    hadHandshake = false;
+    resetClientData();
+  }
+
+  auto Server::reset() -> void {
+    hooks.read.reset();
+    hooks.write.reset();
+    hooks.normalizeAddress.reset();
+    hooks.regReadGeneral.reset();
+    hooks.regWriteGeneral.reset();
+    hooks.regRead.reset();
+    hooks.regWrite.reset();
+    hooks.emuCacheInvalidate.reset();
+    hooks.targetXML.reset();
+
+    resetClientData();
+  }
+
+  auto Server::resetClientData() -> void {
+    breakpoints.reset();
+    breakpoints.reserve(DEF_BREAKPOINT_SIZE);
+
+    watchpointRead.reset();
+    watchpointRead.reserve(DEF_BREAKPOINT_SIZE);
+
+    watchpointWrite.reset();
+    watchpointWrite.reserve(DEF_BREAKPOINT_SIZE);
+
+    pcOverride.reset();
+    insideCommand = false;
+    cmdBuffer = "";
+    haltSignalSent = false;
+    forceHalt = false;
+    singleStepActive = false;
+
+    currentThreadC = -1;
+    hasActiveClient = false;
+    handshakeDone = false;
+    requestDisconnect = false;
```

Comment:
Missing `nonStopMode = false;`. Even if it doesn't work yet, it's good habit to reset it on a new connection.